x-pack/metricbeat/module/gcp/billing/billing.go

```diff
@@ -315,26 +380,134 @@ func generateEventID(currentDate string, rowItems []bigquery.Value) string {
 // generateQuery returns the query to be used by the BigQuery client to retrieve monthly
 // cost types breakdown.
 func generateQuery(tableName, month, costType string) string {
+	if isDetailedTable(tableName) {
+		return createDetailedQuery(tableName, month, costType)
+	}
+
+	return createStandardQuery(tableName, month, costType)
+}
+
+func createStandardQuery(tableName, month, costType string) string {
 	// The table name is user provided, so it may contains special characters.
 	// In order to allow any character in the table identifier, use the Quoted identifier format.
 	// See https://github.com/elastic/beats/issues/26855
 	// NOTE: is not possible to escape backtics (`) in a multiline string
 	escapedTableName := fmt.Sprintf("`%s`", tableName)
+
+	query := fmt.Sprintf(`
+SELECT
+	invoice.month AS invoice_month,
+	project.id AS project_id,
+	project.name AS project_name,
+	billing_account_id,
+	cost_type,
+	IFNULL(sku.id, '') AS sku_id,
+	IFNULL(sku.description, '') AS sku_description,
+	IFNULL(service.id, '') AS service_id,
+	IFNULL(service.description, '') AS service_description,
+	ARRAY_TO_STRING(ARRAY(
+	    SELECT
+	        CONCAT(t.key, ':', t.value)
+	    FROM
+	        UNNEST(tags) AS t), ',') AS tags_string,
+	(SUM(CAST(cost * 1000000 AS int64)) + SUM(IFNULL((
+			SELECT
+				SUM(CAST(c.amount * 1000000 AS int64))
+			FROM
+				UNNEST(credits) c), 0))) / 1000000 AS total_exact
+FROM
+	%s
+WHERE
+	project.id IS NOT NULL
+	AND invoice.month = '%s'
+	AND cost_type = '%s'
+GROUP BY
+	invoice_month,
+	project_id,
+	project_name,
+	billing_account_id,
+	cost_type,
+	sku_id,
+	sku_description,
+	service_id,
+	service_description,
+	tags_string
+ORDER BY
+	invoice_month ASC,
+	project_id ASC,
+	project_name ASC,
+	billing_account_id ASC,
+	cost_type ASC,
+	sku_id ASC,
+	sku_description ASC,
+	service_id ASC,
+	service_description ASC,
+	tags_string ASC;`,
+		escapedTableName, month, costType)
+
+	return query
+}
+
+func createDetailedQuery(tableName, month, costType string) string {
```

Comment:
Tested the changes, and they LGTM. Converting `row` into a struct is a good improvement.

I suggest adding a comment in the `createDetailedQuery` function to explicitly state that it's the same as  `createStandardQuery` with the addition of the `effective_price` column.